typing/typecore.ml

```diff
@@ -7058,7 +7058,9 @@ and type_comprehension_clauses
 and type_comprehension_clause ~loc ~comprehension_type ~container_type env
   : Extensions.Comprehensions.clause -> _ = function
   | For bindings ->
-      reset_pattern true;
+      (* TODO: fix handling of first-class module patterns so we can pass
+       * "true" here. *)
+      reset_pattern false;
```

Comment:
Didn't you also say you had to move this elsewhere?  Or is that only if it's going to be `true`?

Reply:
I don't think we should do this yet -- I tried making the naive change of moving `reset_pattern` to before the call to `type_pat`, but that led to other tests breaking. In other words, it seems non-trivial.

Reply:
Actually I found that this approach was flawed, as, just as the `in` body could disable `allow_modules` mode, so too can it re-enable it. (This means the "unbound module `M`" error could slip through.) Instead, I set the `allow_modules` ref to false just before checking the comprehension pattern.